src/wire/ipv6/extops.rs

```diff
@@ -0,0 +1,421 @@
+            // Any other option other than Pad1 must have a data length
+            // greater than or equal to the extenion option header length.
+            Err(Error::Truncated)
+        } else if len == field::PAYLOAD {
+            if data[1] == 0x00 {
+                // No option data with a header. E.g. PadN for two bytes.
+                Ok(())
+            } else {
+                // The length of bytes given is the size of the extension
+                // option header, but the option data length value6 bytes
+                // zero.
+                Err(Error::Truncated)
+            }
+        } else if len < (self.data_len() as usize) + field::PAYLOAD {
+            Err(Error::Truncated)
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Consume the packet, returning the underlying buffer.
+    #[inline]
+    pub fn into_inner(self) -> T {
+        self.buffer
+    }
+
+    /// Return the option type field.
+    #[inline]
+    pub fn opt_type(&self) -> ExtOpType {
+        let data = self.buffer.as_ref();
+        ExtOpType::from(data[field::OPT_TYPE])
+    }
+
+    /// Return the total length of the header and data.
+    #[inline]
+    pub fn total_len(&self) -> usize {
+        if self.opt_type() == ExtOpType::Pad1 {
+            // If Pad1 Option the size is only one byte
+            field::DATA_LEN
+        } else {
+            field::PAYLOAD + (self.data_len() as usize)
+        }
+    }
+
+    /// Return the option data length field.
+    #[inline]
+    pub fn data_len(&self) -> u8 {
+        let data = self.buffer.as_ref();
+        if self.opt_type() != ExtOpType::Pad1 {
+            data[field::DATA_LEN]
+        } else {
+            0
+        }
+    }
+}
+
+impl<'a, T: AsRef<[u8]> + ?Sized> Packet<&'a T> {
+    /// Return a pointer to the payload.
+    pub fn payload(&self) -> &'a [u8] {
+        let data = self.buffer.as_ref();
+        let start = if self.opt_type() != ExtOpType::Pad1 {
+            field::PAYLOAD
+        } else {
+            field::DATA_LEN
+        };
+        let range = start..self.total_len();
+        &data[range]
+    }
+}
+
+impl<T: AsRef<[u8]> + AsMut<[u8]>> Packet<T> {
+    /// Set the option type field.
+    #[inline]
+    pub fn set_opt_type(&mut self, opt_type: ExtOpType) {
+        let data = self.buffer.as_mut();
+        data[field::OPT_TYPE] = opt_type.into();
+    }
+
+    /// Set the option data length field.
+    ///
+    /// # Panics: If the option type is [Pad1] function this is not valid
+    ///
+    /// [Pad1]: #enum.ExtOpType.Pad1
+    #[inline]
+    pub fn set_data_len(&mut self, data_len: u8) {
+        assert!(self.opt_type() != ExtOpType::Pad1);
+        let data = self.buffer.as_mut();
+        data[field::DATA_LEN] = data_len;
+    }
+}
+
+impl<'a, T: AsRef<[u8]> + AsMut<[u8]> + ?Sized> Packet<&'a mut T> {
+    /// Return a mutable pointer to the payload.
+    pub fn payload_mut(&mut self) -> &mut [u8] {
```

Comment:
Using this on a `Pad1` option is more or less a user error since it cannot have a payload. I went back and forth on using an `assert(self.opt_type() != ExtOpType::Pad1)`.